patcher_test.go

```diff
@@ -0,0 +1,155 @@
+package expr_test
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/expr-lang/expr"
+	"github.com/expr-lang/expr/ast"
+	"github.com/expr-lang/expr/internal/testify/assert"
+	"github.com/expr-lang/expr/internal/testify/require"
+)
+
+func TestExpr_named_types(t *testing.T) {
+	type namedIntType int16
+	type namedFloatType float32
+	type namedStringType string
+	env := map[string]any{
+		"namedInt":    func() namedIntType { return 42 },
+		"basicInt":    func() int { return 42 },
+		"namedFloat":  func() namedFloatType { return 3.0 },
+		"basicFloat":  func() float32 { return 3.0 },
+		"namedString": func() namedStringType { return "abc" },
+		"basicString": func() string { return "abc" },
+	}
+
+	cases := []struct {
+		code     string
+		expected bool
+	}{
+		{`namedFloat() == 3.0`, true},
+		{`namedInt() == 42`, true},
+		{`basicInt() == 42`, true},
+		{`basicFloat() == 3.0`, true},
+		{`namedString() == "abc"`, true},
+		{`basicString() == "abc"`, true},
+		{`namedInt() == 42 && namedFloat() == 3.0 && namedString() == "abc"`, true},
+	}
+	for _, c := range cases {
+		program, err := expr.Compile(c.code,
+			expr.Env(env),
+			expr.Function("unpack", unpack),
+			// expr.Patch(namedTypesPatcher{}),
```

Comment:
why not using `namedTypesPatcher ` instead? It looks more type-safe and there will be no need to call any extra "unpack" function

Reply:
that only worked with `int()` because of that logic here: https://github.com/expr-lang/expr/blob/5e660e753e47a4f3a3a9e234f625dc210b5a331e/builtin/lib.go#L211-L217

but with `float()` it failed, because this type unpacking isn't implemented here: https://github.com/expr-lang/expr/blob/5e660e753e47a4f3a3a9e234f625dc210b5a331e/builtin/lib.go#L252-L255

So I went with one custom function `unpack()` that works with all types (even bool which isn't in these built in functions)